File: examples/pair_consensus_scores.py
from __future__ import annotations

from statistics import fmean

import raven

import _helpers as h


# Raven parameters:
CORESET_SIZE = 256
SAMPLING_SEEDS = 64
NUM_TRIALS = 5
RAVEN_RNG_SEED = 42


# Workload parameters:
WORKLOAD_SEED = 42


workload = h.make_workload(seed=WORKLOAD_SEED)

# initialize Raven index:
index = raven.Raven(
    h.NUM_CLUSTERS,
    coreset_size=CORESET_SIZE,
    sampling_seeds=SAMPLING_SEEDS,
    num_trials=NUM_TRIALS,
    rng_seed=RAVEN_RNG_SEED,
    **h.INDEX_KWARGS,
)
roc_aucs: list[float] = []
avg_precisions: list[float] = []

for batch_index, batch in enumerate(workload.batches):

    # edge_ops is a list of (u, v, weight) tuples, where weight=None indicates deletion
    index.update_edges(batch.edge_ops)

    # flush the index to apply all updates before querying
    # (triggers on query anyway)
    index.flush()

    # get info about the current snapshot
    # including which nodes the workload wants to query + their true labels
    snapshot = h.query_snapshot(
        index,
        workload,
        batch_index,
        batch,
        coreset_size=CORESET_SIZE,
        workload_seed=WORKLOAD_SEED,
    )
    if snapshot is None:
        continue

    # generate balanced same-cluster/cross-cluster node pairs for this snapshot
    pairs, pair_labels = h.balanced_pair_batch(
        snapshot,
        workload.cluster_labels,
        workload_seed=WORKLOAD_SEED,
    )

    # query the unique nodes in those pairs and score each pair using trial consensus
    pair_scores = index.score_pairs(pairs)
    roc_auc, avg_precision = h.pair_score_metrics(pair_labels, pair_scores)
    roc_aucs.append(roc_auc)
    avg_precisions.append(avg_precision)
    print(
        f"batch={snapshot.batch_index:02d} "
        f"time={snapshot.batch_time} "
        f"updates={len(batch.edge_ops)} "
        f"pairs={h.PAIR_COUNT} "
        f"roc_auc={roc_auc:.3f} "
        f"avg_precision={avg_precision:.3f}"
    )

print(f"mean ROC-AUC: {fmean(roc_aucs):.3f}")
print(f"mean average precision: {fmean(avg_precisions):.3f}")

File: crates/raven/examples/pair_consensus_scores.rs
use raven::{EdgeUpdate, Raven, RavenConfig, TrialWeighting};

fn main() -> Result<(), raven::RavenError> {
    let mut config = RavenConfig::new(2);
    config.coreset_size = 4;
    config.sampling_seeds = 2;
    config.num_trials = 5;
    config.rng_seed = Some(42);
    config.node_capacity = 16;
    config.expected_edges_per_node = 4;

    let mut index = Raven::new(config)?;
    index.update_edges([
        EdgeUpdate::set(1, 2, 1.0),
        EdgeUpdate::set(2, 3, 1.0),
        EdgeUpdate::set(1, 3, 1.0),
        EdgeUpdate::set(10, 11, 1.0),
        EdgeUpdate::set(11, 12, 1.0),
        EdgeUpdate::set(10, 12, 1.0),
    ])?;
    index.flush()?;

    let pairs = [(1, 2), (1, 10), (10, 11)];
    let scores = index.score_pairs(&pairs, TrialWeighting::ScoreSoftmax, None)?;

    for ((u, v), score) in pairs.into_iter().zip(scores) {
        println!("pair=({u}, {v}) score={score:.3}");
    }

    Ok(())
}
